be/src/storage/lake/compaction_policy.cpp

```diff
@@ -37,6 +39,93 @@ class BaseAndCumulativeCompactionPolicy : public CompactionPolicy {
+    ~PrimaryCompactionPolicy() override = default;
+
+    StatusOr<std::vector<RowsetPtr>> pick_rowsets(int64_t version) override;
+
+private:
+    double calc_compaction_score(const RowsetStat& stats);
+    static const size_t compaction_result_bytes_threashold = 1000000000;
+    static const size_t compaction_result_rows_threashold = 10000000;
+
+private:
+    TabletPtr _tablet;
+};
+
+StatusOr<std::vector<RowsetPtr>> PrimaryCompactionPolicy::pick_rowsets(int64_t version) {
+    ASSIGN_OR_RETURN(auto tablet_metadata, _tablet->get_metadata(version));
+
+    std::vector<RowsetPtr> input_rowsets;
+    // The goal of lake primary table compaction:
+    // 1. clean up deleted bytes.
+    // 2. merge small rowsets to bigger rowset.
+    // so we pick rowset to compact by this logic:
+    // First, pick out rowset with more deleted bytes.
+    // Second, pick out rowset with less bytes.
+    struct RowsetCompare {
+        static double calc_del_bytes(const RowsetCandidate& rc) {
+            return (double)rc.second.bytes * (double)rc.second.num_dels / (double)rc.second.num_rows;
+        }
+        bool operator()(const RowsetCandidate& a, const RowsetCandidate& b) const {
+            double delete_bytes_a = calc_del_bytes(a);
+            double delete_bytes_b = calc_del_bytes(b);
+            if (delete_bytes_a < delete_bytes_b) {
+                return true;
+            } else if (delete_bytes_a > delete_bytes_b) {
+                return false;
+            } else {
+                // may happen when deleted rows is zero
+                return a.second.bytes > b.second.bytes;
+            }
+        }
+    };
+    UpdateManager* mgr = _tablet->update_mgr();
+    std::priority_queue<RowsetCandidate, std::vector<RowsetCandidate>, RowsetCompare> rowset_queue;
+    for (const auto& rowset_pb : tablet_metadata->rowsets()) {
+        RowsetStat stat;
+        stat.num_rows = rowset_pb.num_rows();
+        stat.bytes = rowset_pb.data_size();
+        stat.num_dels = mgr->get_rowset_num_deletes(_tablet->id(), version, rowset_pb);
+        rowset_queue.push(std::make_pair(std::make_shared<RowsetMetadata>(rowset_pb), stat));
+    }
+    size_t cur_compaction_result_bytes = 0;
+    size_t cur_compaction_result_rows = 0;
+    int i = 0;
+    std::stringstream input_infos;
+    while (!rowset_queue.empty()) {
+        const auto& rowset_candidate = rowset_queue.top();
+        cur_compaction_result_bytes += rowset_candidate.second.bytes;
+        cur_compaction_result_rows += rowset_candidate.second.num_rows;
+        if (input_rowsets.size() > 0 && ((cur_compaction_result_bytes > compaction_result_bytes_threashold * 3 / 2) ||
+                                         (cur_compaction_result_rows > compaction_result_rows_threashold * 3 / 2))) {
+            break;
+        }
+        input_rowsets.emplace_back(std::make_shared<Rowset>(_tablet.get(), std::move(rowset_candidate.first), i++));
```

Comment:
Pass `i` as the rowset index may cause unexpected behaviors.

Reply:
Done